File: tracers/python/av_plain.py
"""Runtime chế độ dán code thuần cho Python.

Python có hook lúc chạy nên không cần biến đổi mã nguồn như Java: `sys.settrace`
cho biết mỗi dòng vừa chạy và toàn bộ biến cục bộ tại đó.

Chỉ in ra bước; phần suy ra vẽ gì nằm ở @av/autoviz, dùng chung cho mọi ngôn ngữ.
"""

import json
import sys

_PREFIX = chr(0x1e) + "@AV|"

# Trần số bước: thuật toán nặng có thể sinh hàng triệu bước và làm nghẽn stdout
MAX_STEPS = 20000
_steps = 0

# Chỉ theo dõi file của người dùng, bỏ qua thư viện chuẩn
_TARGET = "main.py"


def _render(value):
    if isinstance(value, bool):
        return value
    if isinstance(value, int):
        return value if abs(value) <= 9007199254740991 else str(value)
    if isinstance(value, float):
        if value != value:
            return {"$num": "NaN"}
        if value in (float("inf"), float("-inf")):
            return {"$num": "Infinity" if value > 0 else "-Infinity"}
        return value
    if isinstance(value, (list, tuple)):
        return [_render(item) for item in value]
    if isinstance(value, str):
        return value
    return None


def _emit(line, variables):
    payload = {}
    for name, value in variables.items():
        if name.startswith("_"):
            continue
        rendered = _render(value)
        if rendered is not None:
            payload[name] = rendered
    sys.stdout.write(
        _PREFIX + json.dumps({"line": line, "vars": payload}, ensure_ascii=False,
                             separators=(",", ":")) + "\n"
    )


def _tracer(frame, event, _arg):
    global _steps
    if event != "line":
        return _tracer
    if _TARGET not in frame.f_code.co_filename:
        return _tracer
    if _steps >= MAX_STEPS:
        return None
    _steps += 1
    _emit(frame.f_lineno, frame.f_locals)
    return _tracer


def start():
    sys.settrace(_tracer)


def stop():
    sys.settrace(None)

File: tracers/ruby/av_plain.rb
# Runtime chế độ tự trực quan hóa cho Ruby.
#
# TracePoint là hook sạch nhất trong cả nhóm: nó cho biết mỗi dòng vừa chạy và toàn bộ
# biến cục bộ tại đó, mà không phải sửa một ký tự nào của code người dùng.
#
# Chỉ in ra bước; phần suy ra vẽ gì nằm ở @av/autoviz, dùng chung cho mọi ngôn ngữ.

require 'json'

module AvPlain
  PREFIX = 30.chr + '@AV|'

  # Trần số bước: thuật toán nặng có thể sinh hàng triệu bước và làm nghẽn stdout
  MAX_STEPS = 20_000

  @steps = 0
  @trace = nil

  class << self
    def render(value)
      case value
      when Integer then value.abs <= 9_007_199_254_740_991 ? value : value.to_s
      when Float
        return { '$num' => 'NaN' } if value.nan?
        return { '$num' => value.positive? ? 'Infinity' : '-Infinity' } if value.infinite?
        value
      when String, TrueClass, FalseClass, NilClass then value
      when Array then value.map { |item| render(item) }
      end
    end

    def start
      @trace = TracePoint.new(:line) do |tp|
        next unless tp.path.end_with?('main.rb')
        next if @steps >= MAX_STEPS

        @steps += 1
        binding_at = tp.binding
        vars = {}
        binding_at.local_variables.each do |name|
          next if name.to_s.start_with?('_')

          rendered = render(binding_at.local_variable_get(name))
          vars[name.to_s] = rendered unless rendered.nil?
        end
        $stdout.puts PREFIX + JSON.generate({ 'line' => tp.lineno, 'vars' => vars })
      end
      @trace.enable
    end

    def stop
      @trace&.disable
    end
  end
end
